### PR TITLE
doc/md: initial documentation for the extension API

### LINKED ISSUE #1876: Document Extension API
The [Extension API](https://github.com/ent/ent/blob/a19a89a141cf1a5e1b38c93d7898f218a1f86c94/entc/entc.go#L197-L220) was [introduced](https://github.com/ent/ent/commit/52663a4df11c353576142c4ef4cdf5474dad237f) something like two months ago. We keep mentioning it in the blog, but don't have an "official" doc on it in the doc site.

### PR DESCRIPTION
**Pull request by rotemtam**
closes #1876

### CODE REVIEW

File: doc/md/extension-api.md [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+code generation hooks with your extension, implement the `Hooks` method:
+
+```go
+func (s *GreetExtension) Hooks() []gen.Hook {
+    return []gen.Hook{
+        DisallowTypeName("Shalom"),
+    }
+}
+
+// DisallowTypeName ensures a type of name "name" doesn not exist.
```
Comment (masseelch) [inline doc/md/extension-api.md:142]: does not

**Comment by masseelch** [comment]
I think `bundles` is correct, isn't it?

**Comment by rotemtam** [comment]
> I think `bundles` is correct, isn't it?

nope, the _extensions_ (plural) bundle functionality.

File: doc/website/sidebars.js [hunk 2]
```diff
@@ -35,6 +35,7 @@ module.exports = {
         'predicates',
         'aggregate',
         'paging',
+        'extension-api'
```
Comment (a8m) [inline doc/website/sidebars.js:38]: The current website structure uses the "Code Generation" section for documenting the generated API. I don't like it, but I prefer the new extension doc will reside close to the "Templates" page (on "Misc").

We already agreed the website needs a new structure, so we can change it in the future.

File: doc/md/extension-api.md [hunk 3]
```diff
@@ -0,0 +1,186 @@
+---
+id: extension-api
+title: Extension API
+---
+
+### Introduction
+
+The Ent [Extension API](https://github.com/ent/ent/blob/a19a89a141cf1a5e1b38c93d7898f218a1f86c94/entc/entc.go#L197)
```
Comment (a8m) [inline doc/md/extension-api.md:8]:
```suggestion
The Ent [Extension API](https://pkg.go.dev/entgo.io/ent/entc#Extension)
```

File: doc/md/extension-api.md [hunk 4]
```diff
@@ -0,0 +1,186 @@
+---
+id: extension-api
+title: Extension API
```
Comment (a8m) [inline doc/md/extension-api.md:3]:
```suggestion
title: Extensions
```

File: doc/md/extension-api.md [hunk 5]
```diff
@@ -0,0 +1,186 @@
+---
+id: extension-api
```
Comment (a8m) [inline doc/md/extension-api.md:2]:
```suggestion
id: extensions
```

File: doc/md/extension-api.md [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+title: Extension API
+---
+
+### Introduction
+
+The Ent [Extension API](https://github.com/ent/ent/blob/a19a89a141cf1a5e1b38c93d7898f218a1f86c94/entc/entc.go#L197)
+facilitates that creation of code-generation extensions that bundle together [Code-gen Hooks](code-gen.md#code-generation-hooks),
```
Comment (a8m) [inline doc/md/extension-api.md:9]:
```suggestion
facilitates that creation of code-generation extensions that bundle together [codegen hooks](code-gen.md#code-generation-hooks),
```

File: doc/md/extension-api.md [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+[Templates](templates.md) and [Annotations](templates.md#annotations) to create reusable components
+that add new rich functionality to Ent's core.
```
Comment (a8m) [inline doc/md/extension-api.md:11]: Maybe reference the entgql extension as an example?  https://pkg.go.dev/entgo.io/contrib/entgql#Extension

File: doc/md/extension-api.md [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+// GreetExtension implements entc.Extension.
+type GreetExtension {
+	entc.DefaultExtension
+}
+```
+
+### Adding Templates
+
+Ent supports adding [external templates](templates.md) that will be rendered during
+code generation.  To bundle such external templates on an extension, implement the `Templates`
```
Comment (a8m) [inline doc/md/extension-api.md:57]:
```suggestion
code generation. To bundle such external templates on an extension, implement the `Templates`
```

File: doc/md/extension-api.md [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+method:
+```gotemplate title="templates/greet.tmpl"
+{{ define "greet" }}
```
Comment (a8m) [inline doc/md/extension-api.md:60]: Maybe add this:

```suggestion
{{/* Tell Intellij/GoLand to enable the autocompletion based on the *gen.Graph type. */}}
{{/* gotype: entgo.io/ent/entc/gen.Graph */}}

{{ define "greet" }}
```

File: doc/md/extension-api.md [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+```go
+func (*GreetExtension) Templates() []*gen.Template {
+    greet, err := gen.NewTemplate("greet").
+    ParseFiles("templates/greet.tmpl")
+    if err != nil {
+        log.Fatal("entc: failed parsing greet template")
+    }
+    return []*gen.Template{
+        greet,
+    }
```
Comment (a8m) [inline doc/md/extension-api.md:85]:
```suggestion
	return []*gen.Template{
		gen.MustParse(gen.NewTemplate("greet").ParseFiles("templates/greet.tmpl")),
	}
```

File: doc/md/extension-api.md [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+to provide users with the ability to configure the greeting word in the generated
+code:
+
+```go
+// GreetingWord implements entc.Annotation
+type GreetingWord string
+
+func (GreetingWord) Name() string {
+	return "GreetingWord"
+}
```
Comment (a8m) [inline doc/md/extension-api.md:103]:
```suggestion
// GreetingWord implements entc.Annotation.
type GreetingWord string

// Name of the annotation. Used by the codegen templates.
func (GreetingWord) Name() string {
	return "GreetingWord"
}
```

File: doc/md/extension-api.md [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+Now, from within your templates you can access the `GreetingWord` annotation:
+```gotemplate
+func ({{ $receiver }} *{{ $n.Name }}) Greet() string {
+    return "{{ $.Annotations.GreetingWord }}, {{ $n.Name }}"
+}
+```
+
+### Adding Hooks
+
+The entc package provides an option to add a list of hooks (middlewares) to the
```
Comment (a8m) [inline doc/md/extension-api.md:130]: Maybe another link to the codegen hooks section?

File: doc/md/extension-api.md [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+code generation hooks with your extension, implement the `Hooks` method:
+
+```go
+func (s *GreetExtension) Hooks() []gen.Hook {
+    return []gen.Hook{
+        DisallowTypeName("Shalom"),
+    }
+}
+
+// DisallowTypeName ensures a type of name "name" does not exist.
```
Comment (a8m) [inline doc/md/extension-api.md:142]: Maybe:
```suggestion
// DisallowTypeName ensures there is no ent.Schema with the given name in the graph.
```

File: doc/md/extension-api.md [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+			for _, node := range g.Nodes {
+				if node.Name == name {
+					return fmt.Errorf("entc: validation failed, type named %q not allowed.", name)
+				}
+			}
+			return next.Generate(g)
+		})
+	}
+}
+```
```
Comment (a8m) [inline doc/md/extension-api.md:155]: 😍

File: doc/md/extension-api.md [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+
+### Using an Extension in Code Generation
+
+To use an extension in our code-generation configuration, use `entc.Extensions`, a helper
+method that returns an `entc.Option` that applies our chosen extensions:
+
+```go title=entc.go
```
Comment (a8m) [inline doc/md/extension-api.md:162]:
```suggestion
```go title=ent/entc.go
```

**Review by a8m — COMMENTED** [review]
Love this @rotemtam 🚀

File: doc/website/sidebars.js (hunk 2 above)
Comment (rotemtam) [reply doc/website/sidebars.js:38]: kk

**Pull request merged**